2011-07-04  Jakub Jelinek  <jakub@redhat.com>

	PR tree-optimization/49618
	* tree-eh.c (tree_could_trap_p) <case FUNCTION_DECL, case VAR_DECL>:
	For DECL_WEAK decls return true.

--- gcc/tree-eh.c
+++ gcc/tree-eh.c
@@ -2459,6 +2459,13 @@ tree_could_trap_p (tree expr)
 	return true;
       return false;
 
+    case VAR_DECL:
+    case FUNCTION_DECL:
+      /* Assume that accesses to weak vars or functions may trap.  */
+      if (DECL_WEAK (expr))
+        return true;
+      return false;
+
     default:
       return false;
     }
